parser/src/aspects/expr_list.rs

```diff
@@ -0,0 +1,102 @@
+use crate::err::ParseErrorKind::{Expected, Unexpected};
+use crate::moves::{blanks, eod, lookahead, of_type, MoveOperations};
+use crate::parser::{ParseResult, Parser};
+use lexer::token::TokenType;
+use lexer::token::TokenType::Comma;
+
+///An aspect to parse expression lists
+pub(super) trait ExpressionListAspect<'a> {
+    ///Implicit lists are whether A, (A), (A, B, ...) or () (if it can be empty)
+    /// according that `(` and `)` are the [`start`]/[`end`] of the list expression.
+    fn parse_implicit_list<E, F>(
+        &mut self,
+        start: TokenType,
+        end: TokenType,
+        non_empty: bool,
+        parse_element: F,
+    ) -> ParseResult<Vec<E>>
+    where
+        F: FnMut(&mut Self) -> ParseResult<E>;
+
+    ///Explicits lists are whether (A), (A, B, ...) or () (if it can be empty)
+    /// according that `(` and `)` are the start/end of the list expression.
+    /// An explicit list cannot consider an expression A as valid, the expression must be surrounded
+    /// with [`start`] and [`end`]
+    fn parse_explicit_list<E, F>(
+        &mut self,
+        start: TokenType,
+        end: TokenType,
+        non_empty: bool,
+        parse_element: F,
+    ) -> ParseResult<Vec<E>>
+    where
+        F: FnMut(&mut Self) -> ParseResult<E>;
+}
+
+impl<'a> ExpressionListAspect<'a> for Parser<'a> {
+    fn parse_implicit_list<E, F>(
+        &mut self,
+        start: TokenType,
+        end: TokenType,
+        non_empty: bool,
+        mut parse_element: F,
+    ) -> ParseResult<Vec<E>>
+    where
+        F: FnMut(&mut Self) -> ParseResult<E>,
+    {
+        if self.cursor.lookahead(of_type(start)).is_some() {
+            self.parse_explicit_list(start, end, non_empty, parse_element)
+        } else {
+            Ok(vec![parse_element(self)?])
+        }
+    }
+
+    fn parse_explicit_list<E, F>(
+        &mut self,
+        start: TokenType,
+        end: TokenType,
+        non_empty: bool,
```

Comment:
I don't think that an "is empty" check should be done by this method, because the error message depends on the method caller.

Reply:
yes indeed